### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry): add `resLE`

### PR DESCRIPTION
**Pull request by chrisflav**
Adds a shortcut `Scheme.Hom.resLE` for the restriction of a morphism of schemes to opens on the source and the target.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/Restrict.lean [hunk 1]
```diff
@@ -216,8 +216,8 @@ def Scheme.restrictFunctor : X.Opens ⥤ Over X where
 @[simp] lemma Scheme.restrictFunctor_obj_hom (U : X.Opens) :
   (X.restrictFunctor.obj U).hom = U.ι := rfl
 
-@[simp] lemma Scheme.restrictFunctor_map_left {U V : X.Opens} (i : U ⟶ V) :
-  (X.restrictFunctor.map i).left = IsOpenImmersion.lift (V.ι) U.ι (by simpa using i.le) := rfl
+lemma Scheme.restrictFunctor_map_left {U V : X.Opens} (i : U ⟶ V) :
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Restrict.lean:219]: Can you add a comment here saying that it is deliberately not a simp lemma?

File: Mathlib/AlgebraicGeometry/Restrict.lean [hunk 2]
```diff
@@ -515,6 +515,68 @@ instance {X Y : Scheme.{u}} (f : X ⟶ Y) (U : Y.Opens) [IsOpenImmersion f] :
...
+lemma map_resLE (i : V' ⟶ V) :
+    (X.restrictFunctor.map i).left ≫ f.resLE U V e = f.resLE U V' (i.le.trans e) := by
+  simp only [Scheme.Hom.resLE, homOfLE_leOfHom, ← Over.comp_left_assoc, ← Functor.map_comp]
+  rfl
+
+/-- Variant of `map_resLE` for equality. -/
+@[reassoc]
+lemma map_resLE' (i : V' = V) :
+    (X.restrictFunctor.map (eqToHom i)).left ≫ f.resLE U V e = f.resLE U V' (i ▸ e) :=
+  map_resLE _ _ _
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Restrict.lean:545]: Do we need this?

**Review by erdOne — COMMENTED** [review]
Thanks!
Also do you think `resLE` should be the new simpNF? Or maybe I imagine it would be useful to curate a simp set where the simpNF is `resLE` and `appLE` and turn it on optionally when needed.

File: Mathlib/AlgebraicGeometry/Restrict.lean (hunk 2 above)
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Restrict.lean:545]: I don't know, I just copied it from the `appLE` case. Removed it for now.

File: Mathlib/AlgebraicGeometry/Restrict.lean [hunk 3]
```diff
@@ -515,6 +515,68 @@ instance {X Y : Scheme.{u}} (f : X ⟶ Y) (U : Y.Opens) [IsOpenImmersion f] :
...
+    Arrow.mk ((f.resLE U V e).app ⊤) ≅ Arrow.mk (f.appLE U V e) :=
+  Arrow.isoMk U.topIso V.topIso <| by
+  simp only [Scheme.Hom.resLE, homOfLE_leOfHom, Scheme.restrictFunctor_map_left,
+    Scheme.comp_app, Arrow.mk_left, Arrow.mk_right,
+    Scheme.Opens.topIso_hom, Arrow.mk_hom, Scheme.Hom.map_appLE, morphismRestrict_app',
+    IsOpenImmersion.lift_app, Scheme.Opens.ι_appIso, Scheme.Opens.ι_app,
+    Scheme.Opens.toScheme_presheaf_map, Category.assoc, ← X.presheaf.map_comp]
+  erw [Category.id_comp]
+  simp only [Scheme.Hom.appLE, Category.assoc, ← X.presheaf.map_comp]
+  rfl
```
[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Restrict.lean:578]: Turns out it is not too bad.

**Comment by chrisflav** [comment]
> Thanks! Also do you think `resLE` should be the new simpNF? Or maybe I imagine it would be useful to curate a simp set where the simpNF is `resLE` and `appLE` and turn it on optionally when needed.

I don't know about making `resLE` the "default" simpNF. But curating a `simp` set for `resLE` and `appLE` sounds reasonable. Also I am not sure if the underlying morphism of `resLE` currently appears very often? So maybe we should see how it is used and then make a follow up PR?

File: Mathlib/AlgebraicGeometry/Restrict.lean [hunk 4]
```diff
@@ -515,6 +521,81 @@ instance {X Y : Scheme.{u}} (f : X ⟶ Y) (U : Y.Opens) [IsOpenImmersion f] :
...
+  rw [← preimage_comp, ← resLE_comp_ι f e, preimage_comp, preimage_image_eq]
+
+lemma le_preimage_resLE_iff {U : Y.Opens} {V : X.Opens} (e : V ≤ f ⁻¹ᵁ U)
+    (O : U.toScheme.Opens) (W : V.toScheme.Opens) :
+    W ≤ (f.resLE U V e) ⁻¹ᵁ O ↔ V.ι ''ᵁ W ≤ f ⁻¹ᵁ U.ι ''ᵁ O := by
+  simp [resLE_preimage, ← image_le_image_iff V.ι, image_preimage_eq_opensRange_inter, V.ι_image_le]
+
+lemma resLE_appLE {U : Y.Opens} {V : X.Opens} (e : V ≤ f ⁻¹ᵁ U)
+    (O : U.toScheme.Opens) (W : V.toScheme.Opens) (e' : V.ι ''ᵁ W ≤ f ⁻¹ᵁ U.ι ''ᵁ O) :
+    (f.resLE U V e).appLE O W ((le_preimage_resLE_iff f e O W).mpr e') =
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Restrict.lean:578]: I think

File: Mathlib/AlgebraicGeometry/Restrict.lean [hunk 5]
```diff
@@ -515,6 +521,81 @@ instance {X Y : Scheme.{u}} (f : X ⟶ Y) (U : Y.Opens) [IsOpenImmersion f] :
...
+      f.appLE (U.ι ''ᵁ O) (V.ι ''ᵁ W) e' := by
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Restrict.lean:579]:
```suggestion
lemma resLE_appLE {U : Y.Opens} {V : X.Opens} (e : V ≤ f ⁻¹ᵁ U)
    (O : U.toScheme.Opens) (W : V.toScheme.Opens) (e') :
    (f.resLE U V e).appLE O W e' =
      f.appLE (U.ι ''ᵁ O) (V.ι ''ᵁ W) ((le_preimage_resLE_iff f e O W).mp e') := by
```
I think this is more useful for rewriting. (Not tested)

**Review by erdOne — COMMENTED** [review]
Otherwise LGTM

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**